pkg/controllers/oauthendpoints/oauth_endpoints_controller.go

```diff
@@ -0,0 +1,273 @@
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	routev1 "github.com/openshift/api/route/v1"
+	configv1informers "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	configv1lister "github.com/openshift/client-go/config/listers/config/v1"
+	routev1informers "github.com/openshift/client-go/route/informers/externalversions/route/v1"
+	routev1listers "github.com/openshift/client-go/route/listers/route/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/libs/endpointaccessible"
+)
+
+// NewOAuthRouteCheckController returns a controller that checks the health of authentication route.
+func NewOAuthRouteCheckController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForTargetNS informers.SharedInformerFactory,
+	kubeInformersForConfigManagedNS informers.SharedInformerFactory,
+	routeInformerNamespaces routev1informers.RouteInformer,
+	ingressInformerAllNamespaces configv1informers.IngressInformer,
+	systemCABundle []byte,
+	recorder events.Recorder,
+) factory.Controller {
+	cmLister := kubeInformersForConfigManagedNS.Core().V1().ConfigMaps().Lister()
+	cmInformer := kubeInformersForConfigManagedNS.Core().V1().ConfigMaps().Informer()
+
+	secretLister := kubeInformersForTargetNS.Core().V1().Secrets().Lister()
+	secretInformer := kubeInformersForTargetNS.Core().V1().Secrets().Informer()
+	routeLister := routeInformerNamespaces.Lister()
+	routeInformer := routeInformerNamespaces.Informer()
+	ingressLister := ingressInformerAllNamespaces.Lister()
+	ingressInformer := ingressInformerAllNamespaces.Informer()
+
+	endpointListFunc := func() ([]string, error) {
+		return listOAuthRoutes(routeLister, recorder)
+	}
+
+	getTLSConfigFunc := func() (*tls.Config, error) {
+		return getOAuthRouteTLSConfig(cmLister, secretLister, ingressLister, systemCABundle, recorder)
+	}
+
+	return endpointaccessible.NewEndpointAccessibleController(
+		"OAuthServerRoute",
+		operatorClient,
+		endpointListFunc, getTLSConfigFunc,
+		[]factory.Informer{
+			cmInformer,
+			secretInformer,
+			routeInformer,
+			ingressInformer,
+		},
+		recorder)
+}
+
+// NewOAuthServiceCheckController returns a controller that checks the health of authentication service.
+func NewOAuthServiceCheckController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForTargetNS informers.SharedInformerFactory,
+	recorder events.Recorder,
+) factory.Controller {
+	endpointsListFunc := func() ([]string, error) {
+		return listOAuthServices(kubeInformersForTargetNS.Core().V1().Services().Lister(), recorder)
+	}
+
+	getTLSConfigFunc := func() (*tls.Config, error) {
+		return getOAuthEndpointTLSConfig(kubeInformersForTargetNS.Core().V1().ConfigMaps().Lister(), recorder)
+	}
+
+	return endpointaccessible.NewEndpointAccessibleController(
+		"OAuthServerService",
```

Comment:
why not combine `OAuthServerService` and `OAuthServerServiceEndpoints`?

Reply:
they can be perceived as two different layers